File: 3439. Reschedule Meetings for Maximum Free Time I/reschedule_meetings_for_maximum_free_time_I.py
# Sliding Window
# T.C.->O(n)
# S.C.->O(n)
class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: List[int], endTime: List[int]) -> int:
        freeArray = []
        # Starting gap
        freeArray.append(startTime[0])
        # In-between free gaps
        for i in range(1, len(startTime)):
            freeArray.append(startTime[i] - endTime[i - 1])
        # Ending gap
        freeArray.append(eventTime - endTime[-1])
        # Sliding window of size k+1 over freeArray
        i = 0
        j = 0
        currSum = 0
        maxSum = 0
        n = len(freeArray)
        while j < n:
            currSum += freeArray[j]
            if j - i + 1 > k + 1:
                currSum -= freeArray[i]
                i += 1
            maxSum = max(maxSum, currSum)
            j += 1
        return maxSum



#Prefix-Sum
#T.C.->O(n)
#S.C.->O(n)
class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: List[int], endTime: List[int]) -> int:
        n = len(startTime)
        res = 0
        total = [0] * (n + 1)
        for i in range(n):
            total[i + 1] = total[i] + endTime[i] - startTime[i]
        for i in range(k - 1, n):
            right = eventTime if i == n - 1 else startTime[i + 1]
            left = 0 if i == k - 1 else endTime[i - k]
            res = max(res, right - left - (total[i + 1] - total[i - k + 1]))
        return res

File: 3439. Reschedule Meetings for Maximum Free Time I/reschedule_meetings_for_maximum_free_time_I.cpp
// Sliding Window
// T.C.->O(n)
// S.C.->O(n)
class Solution {
public:
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        vector<int> freeArray; // store durations of free gaps
        //free gap duration=ith start - (i-1)th end
        freeArray.push_back(startTime[0]); //starting gap
      
        for(int i=1; i<startTime.size();i++)
        { freeArray.push_back(startTime[i] - endTime[i-1]); }
      
        freeArray.push_back(eventTime - endTime[endTime.size()-1]); //ending gap
      
        //SLIDING WINDOW
        int i=0;
        int j=0;
        int maxSum=0;
        int currSum=0;
        int n=freeArray.size();
        while(j<n) {
            currSum+=freeArray[j];
            if(i< n && j-i+1 > k+1) { //k+1 is window size
                currSum-=freeArray[i];
                i++;
            }
            maxSum=max(maxSum,currSum);
            j++;
        }
        return maxSum;
    }
};

//Prefix-Sum
//T.C.->O(n)
//S.C.->O(n)
class Solution {
public:
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        int n = startTime.size(), res = 0;
        vector<int> sum(n + 1);
        for (int i = 0; i < n; i++) {
            sum[i + 1] = sum[i] + endTime[i] - startTime[i];
        }
        for (int i = k - 1; i < n; i++) {
            int right = i == n - 1 ? eventTime : startTime[i + 1];
            int left = i == k - 1 ? 0 : endTime[i - k];
            res = max(res, right - left - (sum[i + 1] - sum[i - k + 1]));
        }
        return res;
    }
};
